src/core.c/Exception.pm6

```diff
@@ -428,28 +428,68 @@ my class CX::Last does X::Control {
 my class CX::Take does X::Control {
     method message() { "<take control exception>" }
 }
+my class CX::Return does X::Control {
+    method message() { "<return control exception>" }
+}
+my class CX::Emit does X::Control {
+    method message() { "<emit control exception>" }
+}
+my class CX::Done does X::Control {
+    method message() { "<done control exception>" }
+}
+my class CX::Succeed does X::Control {
+    method message() { "<succeed control exception>" }
+}
+my class CX::Proceed does X::Control {
+    method message() { "<proceed control exception>" }
+}
+
 my class CX::Warn does X::Control {
     has $.message;
 
     method UPGRADE-RAT(Int $nu, Int $de) is raw {
         warn "Downgrading Rat $nu / $de to Num";
         nqp::p6box_n(nqp::div_In($nu,$de))
     }
+
+    method WARN(str $message, $bt --> Nil) is hidden-from-backtrace {
+        note "$message\n$bt.first-none-setting-line()";
+    }
 }
-my class CX::Succeed does X::Control {
-    method message() { "<succeed control exception>" }
-}
-my class CX::Proceed does X::Control {
-    method message() { "<proceed control exception>" }
+my class CX::Warn::Quietly {
```

Comment:
How does `&EXCEPTION` even enter into this?

The `CX::Warn::xxx` classes are not to be thrown.  The reason I chose their name that way, was a. to prevent more global name pollution.  I guess they could well be called `Warnings::Quietly`, `Warnings::Fatal`, etc.  Also, there's a bit of prior art with `$*UPGRADE-RAT` which currently takes the `CX::Warn` class to warn.

To re-iterate: the only time when `$*WARNINGS` is checked, is when a warning has already been thrown, and we're about to print the warning on STDERR.  Only **THEN** does it act differently (potentially) from before.

If your main beef is with the name of the `CX::Warn::xxx`, I can change them

Reply:
> To re-iterate: the only time when `$*WARNINGS` is checked, is when a warning has already been thrown, and we're about to print the warning on STDERR. Only **THEN** does it act differently (potentially) from before.

This is the key point I consider wrong.

Once again, the idea is to let user code decide their behavior depending on the current kind of warning. Best this is done if they can multi-dispatch over warning type object. To have this working correctly, as I stated, all of them must inherit from `CX::Warn` and conversion from VM-level control exception into a HLL `CX::Warn` instance must take place in `EXCEPTION`. In this case all `print_control` would have to do is invoke `WARN` method on a received `CX::Warn` instance. And when done this way the code in my examples would become possible.

Reply:
> conversion from VM-level control exception into a HLL CX::Warn instance must take place in EXCEPTION

It has never done so in the past.  I fail to see why now.

As to:
````raku
CONTROL {
    when CX::Warn::Debug { ... }
    when CX::Warn::Fatal { ...; .rethrow }
}
````
the whole point to *me* was not needing clunky `CONTROL` blocks.  And how would you throw a `CX::Warn::Fatal` in the first place??  What am I missing?

The whole idea was to get some sort of *global* control on the behaviour of `warn` and `CX::Warn` control exceptions thrown in another way.  Similar to how `%*ENV<RAKU_EXCEPTIONS_HANDLER>` allows you to specify how fatal exceptions are to be handled.  But with a lexical dynamic override.  And the possibility of `CONTROL` to still take local control about what to do any given warnings.

The example of the `CONTROL` block doesn't make any sense anyway, as in a given scope, you would only have one behaviour of warnings active anyway (unless you start doing *really* weird things as setting the value of `$*WARNINGS` multiple times in the same scope.

So in short: I'm not getting what your examples would actually add in useful functionality, other than to make them possible?

Reply:
So much to explain, so much has been understood incorrectly! :) 

> > conversion from VM-level control exception into a HLL CX::Warn instance must take place in EXCEPTION
> 
> It has never done so in the past. I fail to see why now.

[Current implementation]{https://github.com/rakudo/rakudo/blob/main/src/core.c/Exception.pm6#L476} does it for `CONTROL` blocks to work correctly. It would need to be slightly extended to support `$*WARNINGS` and that's all.

> the whole point to _me_ was not needing clunky `CONTROL` blocks.

The `CONTROL` example is not about throwing, it is about **handling** thrown exceptions. Once again, step back a little from your producing-centric point of view and try yourself in the shoes of one handling the warnings and wanting to have different reactions to different ones. What would they need to do in your current design? The `CONTROL` example is what they'd need in my case. Notice, please, that `$*WARNINGS` can be overridden (you stated it already) and same application can throw different kinds of warnings. Not necessarily that a warning thrown would be processed right away in the same `CONTROL` block, where it was caught, it could be postponed passed into a different thread or otherwise lose the context – hence, lose access to the original `$*WARNING`.

My approach doesn't require you to use `CONTROL` to throw/re-throw warnings, it preserves the whole thing about having `$*WARNING`. It just lets one to have more elegant, logical, and concise handling code.

> The example of the `CONTROL` block doesn't make any sense anyway, as in a given scope, you would only have one behaviour of warnings active anyway (unless you start doing _really_ weird things as setting the value of `$*WARNINGS` multiple times in the same scope.

Wrong because `CONTROL` may sit very early in the call stack. `$*WARNING` can be changed deep down the call chain, even by a 3rd-party module.

> And how would you throw a `CX::Warn::Fatal` in the first place??

```raku
class W::Foo is CX::Warn {
    method message { "Foo!" }
}

{
    CONTROL {
        note "OK? ", .raku;
        when W::Foo {
            note "It's W::Foo";
            .resume
        }
    }

    my $a = 42;
    W::Foo.new.throw if $a > 30;
    $a = 5;
    say $a;
}
```

This is something I played with a while ago. The only problem this code currently has is that the core doesn't know how to handle `W::Foo` because no one considered a possibility of inheriting from `CX::Warn`. So, replacing `.resume` with `.rethrow` or simply removing the `CONTROL` block results in "control exception without handler". I just haven't got time to take care of it.

> So in short: I'm not getting what your examples would actually add in useful functionality, other than to make them possible?

I hope I stated it above clearly, but worth repeating: simplify **handling** of different kinds of warnings thrown would it be required. Even without considering manipulations with `$*WARNINGS', I'd feel much better having to multi-dispatch over an exception type rather than analyzing what's in a dynamic variable which value isn't even guaranteed to be preserved for the time, when I start processing it.

Reply:
But, a `warn "foo"` doesn't call `EXCEPTION` at all????  Only if you have a `CONTROL { }` block active, does it run `EXCEPTION`?

Reply:
Not sure what you meant there, but that's the reason for *me* to put the check on `$*WARNINGS` in `print_control`.

Now, if you find the `CX::Warn` class naming confusing, I can change that.

Reply:
No, it's not about the naming. It is about **handling** by user code would that be necessary. It is about being able to multi-dispatch within `CONTROL` on a caught warning exception.

Simple, in your implementation `print_control` reads `$*WARNINGS` and interprets it. `EXCEPTION` behavior is unchanged, it creates a `CX::Warn` for user when needed.

I propose to teach `EXCEPTION` to read and interpret `$*WARNINGS`. `print_control` might use this ability to get a `CX::Warn` to call `WARN` upon. This way `$*WARNINGS` interpretation is localized in one place, but the result of it can be used by either user's `CONTROL` or by `print_control` equally well.

Reply:
And, BTW, I don't want them to be renamed. :)